src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PackageDeclarationCheck.java

```diff
@@ -145,11 +145,6 @@ public int[] getAcceptableTokens() {
         return getRequiredTokens();
     }
 
-    @Override
-    public void beginTree(DetailAST ast) {
-        defined = false;
```

Comment:
https://github.com/checkstyle/checkstyle/blob/655f743503209544b1b87dc016da4ebfd10cee12/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PackageDeclarationCheck.java#L103

It means that Check instance will stay same between two input java classes. It has to be reset his state on each new file.

To reproduce problem, handle two files, one with violation and without. There should be violation only on first.

Recheck git history on changes for this field to get more details why it is required.

Reply:
@shubh220922 it is likely that we would need two concurrent executions of check in different threads to kill this mutation, if I understand documentation correctly.

Reply:
@nrmancuso ok I'm looking into it.

Reply:
@nrmancuso 
I studied multi threading as much as possible. I would request you to send link of the documentation you are talking about in https://github.com/checkstyle/checkstyle/pull/12904#discussion_r1163522418 and also point me to any pre existing multi threaded test in checkstyle's codebase as I have not written multi threaded tests before.

Reply:
@nrmancuso @romani please let me know what I can do here.

Reply:
Looks like we have different instances of the check between runs, @shubh220922 you should be able to determine this by checking the hascode of the check object.

Reply:
in real execution Check instance is always the same for all files so this is why cleanup is required.
In test we need to simulate same,  it might require verify method that takes two Input classes for execution, I think we already has such method.

Reply:
@romani could you be more precise. We do have ``verify()`` method that takes 2 input files at once but, it is only for those files that have same violations. Please refer the method below:
```java
/**
     * Performs verification of two files with their given file paths using specified
     * configuration of one file only. Also performs verification of the config specified
     * in the input file. This method needs to be implemented when two given files need to be
     * checked through a single check only.
     *
     * @param filePath1 file path of first file to verify
     * @param filePath2 file path of second file to verify
     * @param expected an array of expected messages
     * @throws Exception if exception occurs during verification process
     */
    protected final void verifyWithInlineConfigParser(String filePath1,
                                                      String filePath2,
                                                      String... expected)
            throws Exception {
        final TestInputConfiguration testInputConfiguration =
                InlineConfigParser.parse(filePath1);
        final DefaultConfiguration parsedConfig =
                testInputConfiguration.createConfiguration();
        verifyViolations(parsedConfig, filePath1, testInputConfiguration.getViolations());
        verifyViolations(parsedConfig, filePath2, testInputConfiguration.getViolations());
        verify(createChecker(parsedConfig),
                new File[] {new File(filePath1), new File(filePath2)},
                filePath1,
                expected);
    }
```
The method: 
```java
verify(createChecker(parsedConfig),
                new File[] {new File(filePath1), new File(filePath2)},
                filePath1,
                expected);
```
takes only one violation.

Reply:
>We do have verify() method that takes 2 input files at once but, it is only for those files that have same violations. 

Nothing in doc of it and implementation mention it.
It takes config from first input file, and validate both files as one run. 
Make two inputs with same inlined config. And use this method

Reply:
Second file should make `ast` as null
https://github.com/checkstyle/checkstyle/blob/70946a8f01ae95c5debf3f7ea10557daadea1cd6/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PackageDeclarationCheck.java#L155

Please investigate if this possible.